pkgs/build-support/rust/default.nix

```diff
@@ -114,24 +114,33 @@ stdenv.mkDerivation (filteredArgs // {
     EOF
 
     export RUST_LOG=${logLevel}
-  '' + stdenv.lib.optionalString validateCargoDeps ''
-    if ! diff source/Cargo.lock $cargoDepsCopy/Cargo.lock ; then
-      echo
-      echo "ERROR: cargoSha256 is out of date"
-      echo
-      echo "Cargo.lock is not the same in $cargoDepsCopy"
-      echo
-      echo "To fix the issue:"
-      echo '1. Use "1111111111111111111111111111111111111111111111111111" as the cargoSha256 value'
-      echo "2. Build the derivation and wait it to fail with a hash mismatch"
-      echo "3. Copy the 'got: sha256:' value back into the cargoSha256 field"
-      echo
-
-      exit 1
-    fi
+  '' + (args.postUnpack or "");
+
+  # After unpacking and applying patches, check that the Cargo.lock matches our
+  # src package. Note that we do this after the patchPhase, because the
+  # patchPhase may create the Cargo.lock if upstream has not shipped one.
+  postPatch = (args.postPatch or "") + stdenv.lib.optionalString validateCargoDeps ''
+    (
+      cd ..
+      echo "Validating Cargo.lock"
+      if ! diff source/Cargo.lock $cargoDepsCopy/Cargo.lock; then
+        echo
+        echo "ERROR: cargoSha256 is out of date"
+        echo
+        echo "Cargo.lock is not the same in $cargoDepsCopy"
+        echo
+        echo "To fix the issue:"
+        echo '1. Use "1111111111111111111111111111111111111111111111111111" as the cargoSha256 value'
+        echo "2. Build the derivation and wait it to fail with a hash mismatch"
+        echo "3. Copy the 'got: sha256:' value back into the cargoSha256 field"
+        echo
+
+        exit 1
+      fi
+    )
   '' + ''
     unset cargoDepsCopy
-  '' + (args.postUnpack or "");
+  '';
```

Comment:
There are ~20 Rust applications that need this before they can use the newer fetcher.  They're working just fine on the old fetcher for the moment, since they were never validating anything (the old validate flag defaulted to false and only a few packages had opted in).  

Of course, it's debatable how valuable validation is when *we* are the ones providing the Cargo.lock in the first place, but this mainly serves to keep the branching/conditionals in the code limited.